Add GEP-1364 proposal

youngnick opened the pull request:
**What type of PR is this?**
/kind gep

**What this PR does / why we need it**:
This PR adds an implementable proposal for GEP1364 (#1364).

It's a big GEP, for a _very_ big change, that will definitely break conformance tests. But I think that it will set us up with clear rules for Conditions going forward.

It's quite opinionated, so I expect to need to defend my decisions here somewhat, don't be shy about objections.

**Which issue(s) this PR fixes**:
Updates #1364 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note
Status definitions have been updated across the API.
```
The release note will need something like the tl;dr in the GEP, once we all agree. 😄

howardjohn commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+is no missing information, including but not limited to:
+  * Any mandatory references resolve to existing resources (examples here are the
+  Gateway's gatewayClass field, or the `parentRefs` field in Route resources)
+  * Any specified TLS secrets exist
```
Currently the API supports partially-valid resources. I believe this is proposing a partially-valid resource would be `Attached=false`, but the config is actually 'active' (or, I would describe it, "attached").

To me this feels a bit odd. The old API sort of had this issue as well.

Thinking purely from a pedantic-correctness POV (which may not be ideal - sometimes simpler is better), I would expect there to be two conditions (I use bad names intentionally to avoid confusion): "ThisResourceWasApplied" - any part of the rule had some effect on the system, and "ThisRuleIsEntirelyValid" - no errors found in the rule at all.

Its possible my concern would be addressed with a different word than "Attached", rather than more conditions, as well

WDYT?

howardjohn replied:
I guess this is somewhat addressed below.. but I am not sure its accurate? It seems we are defining a very small set of "failures that do not become unattached", but there are a lot more partially-accepted resources I think?

youngnick replied:
The guideline I've been basing the "does something become unattached?" on is "Does this produce at least some configuration in the underlying data plane?". So I guess it sounds like I need to be more explicit with that guideline, but that aside from that, we're largely in agreement?

howardjohn replied:
I think it makes sense to have "Is this attached" and "is this 100% valid" as two separate things. 

Part of my confusion, I think, is I am not actually sure which fields result in "not attached". It seems _most_ issues are partially accepted. Which is fine - `Attached` just becomes basically a pretty high level acknowledgement that the controller is handling the resource

keithmattix replied:
Yeah I think I agree with @howardjohn's point here: "does this resource produce config" is distinct from "does this resource produce _valid_ config". I'm sure some implementations don't allow invalid (e.g. duplicate route names or something) to be sent to the data plane, but it doesn't feel like that's a requirement for the Gateway API at this juncture.

youngnick replied:
I guess I wrote this pooly - by "Does this resource produce config" I meant to read "Is this resource valid enough to produce some config when combined with the rest of the resource set".

The thing I'm thinking of particularly here is that for HTTPRoute, invalid references still produce data path config because the match will respond with 500s. Whereas for TCPRoute etc, if you don't have a service to route to, there's no way you can build a forwarding path - there's nothing to forward to! So that won't produce any data path config at all, and so, in the current design, would cause the resource to fail to be attached.

I think that partial validity is an enormous minefield of weirdness, just like not applying config if it would break things.

evankanderson replied:
Is it possible for something like an HTTPRoute to define two paths, one of which is "valid enough" config, while the other is garbage?

If so, are there expectations or requirements from a conformance PoV ("always apply all or none") that we need to specify?

sunjayBhatia replied:
> Is it possible for something like an HTTPRoute to define two paths, one of which is "valid enough" config, while the other is garbage?

Most cases of an invalid configuration in one route rule (with others being completely valid within the same HTTPRoute) seem to be misconfigurable only in the context of a reference to another resource and fall into the `ResolvedRefs` bucket (which seems stated below and possibly agreed upon that `Attached: true` be set with `ResolvedRefs: false` and return HTTP code 500 for the invalid rule).

One such example not covered is if you were to use a `HTTPRequestRedirectFilter` and `HTTPURLRewrite` in the list of filters for an individual `HTTPRouteRule` ([ref](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRequestRedirectFilter)). At the moment it doesn't appear we have prescribed any status code or other error handling for this error. This seems to fall into the bucket of things that cannot produce data plane configuration.

Going by the statements below, if this rule were the only one present `Attached: false` should be set, since it cannot result in any configuration (with another more specific error condition). If present with other valid rules, `Attached: true` should be set (with another more specific error condition again). I can definitely see the argument for setting `Attached: false` however in the latter case, since stated above, the `Attached` condition means `The resource is syntatically and semantically valid, and internally consistent`.

It seems that the `Valid: true` condition mentioned above ends up being somewhat equivalent to having  for example for HTTPRoute `ResolvedRefs: true` and no error conditions (and vice versa, `Valid: false` equivalent to `ResolvedRefs: false` *or* some more specific error condition). Maybe this framing helps a little with determining possible merits of having another condition?

> If so, are there expectations or requirements from a conformance PoV ("always apply all or none") that we need to specify?

Just doing a quick pass, I don't *think* I saw any existing conformance tests that cover this kind of thing so I agree that would be good to add.

youngnick replied:
Avoiding "always apply all or none" is the reason this is tricky. There are mixed expectations about what happens if _some_ config is present but bad, and I'm attempting to thread the needle here and make this as user-friendly as possible while not becoming really difficult to implement or understand.

youngnick replied:
I added a section on partial validity to hopefully make this clearer.

shaneutt commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+
+## TLDR
+
+The status, particularly the Conditions, across the whole Gateway API has very much grown
+grown organically, and so has many inconsistencies and odd behaviors.
```
```suggestion
The status, particularly the Conditions, across the whole Gateway API has very much grown
organically, and so has many inconsistencies and odd behaviors.
```

robscott replied:
Tiny nit: has -> have

howardjohn commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+  stuck with some sort of "will be ready in no longer than x seconds" standard
+  (which we'll also need to write conformance tests for, probably including a 
+  stress test of adding a *lot* of records as well.) As an initial ballpark,
+  I propose 5 seconds.
```
This probably won't work for cloud providers (at least GKE is slower, not sure about others).

I am also not sure it provides value. What do we lose if we don't define anything here?

If we consider something like Knative, they certainly aren't going to be able to just wait X seconds; I'd expect the same with other users

howardjohn replied:
And if we can't provide any guarantees about Readiness, then it seems like Ready becomes no different than Attached...

youngnick replied:
We certainly can mandate `Ready` to be an _extended_ condition, and say "If you supply a Ready condition, it has to work like this". Maybe doing that is a better place to start with `Ready`?

howardjohn replied:
Do we have a feel for which implementations can and cannot support `Ready` properly?

robscott replied:
Agree that cloud providers will likely have trouble implementing a `Ready` condition. Could we introduce a `Reconciled` condition that would be more broadly implementable?

bowei replied:
+1 any distributed (e.g. geo distributed) implementation has a hard time with this (think: config was pushed to 95% of the sites around the world, is it ready?)

howardjohn reviewed (commented):
Definitely a good direction

keithmattix commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+
+Because this GEP is mainly concerned with updating the Conditions we are setting in
+Gateway API resources' `status`, it's worth reviewing some important points about
+Conditions. (Thi information is mainly taken from the [Typical status properties][typstatus]
```
Missing href for the link here?

youngnick replied:
It's all the way down at the bottom of the doc, with the other references. I thought it was better to have everything in the one place.

keithmattix commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+one entry of each item, using the `type` field as a key. (So, this is effectively
+a map that looks like a list in YAML form).
+* Each has a number of fields, the most important of which for this discussion
+are `type`, `status`, `season`, and `observedGeneration`.
```
```suggestion
are `type`, `status`, `reason`, and `observedGeneration`.
```

keithmattix commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+for processing witll be replaced with `Attached`, except for GatewayClass (since
+it is the root of the resource tree, it will stay with `Accepted`).
+* In general, resources should be considered `Attached` if their config will
+generate some config in the underlying data plane.
```
It might be helpful to be explicit here:
```suggestion
generate some config in the underlying data plane (whether that config is valid or not).
```

youngnick replied:
I've been assuming that only valid config will be passed to the data plane. If a resource would generate invalid config, then to my mind it shouldn't be considered attached.

Again, here I'm thinking of HTTPRoute adding 500s for things that would otherwise cause a resource to become unattached (that is, it doesn't  match anything, or it doesn't have a valid `backendRef`, or there are zero endpoints in a valid backendRef). For Route resources that don't have an error code response like HTTP does, presence of that class of error should probably result in it becoming unattached (since you can't generate any meaningful data plane config at all).

keithmattix commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+
+Note that some classes of inter-resource reference failure do _not_ cause a resource
+to become unattached (that is, to have the `Attached` condition set to `status: false`).
+* Nonexistent Service backends - if the backend does not exist on a HTTPRoute that
```
This starts painting an observability problem IMO.  Even if conditions are only meant to be human-readable, how does a human look at an HTTPRoute resource and know "oops I have a typo in my service backend"? Yes, 500s indicate something's wrong, but I would expect to find a condition pointing to the specific issue

bowei replied:
We should allow room for error conditions in addition to attached in this case so the intformation can be conveyed.

youngnick replied:
Yes, I guess I should make this clearer, that in this case, it's absolutely expected that there be some additional, negative-polarity error conditions that tell you that the backendRef doesn't exist. (This is a distinct case from the backend existing, but having zero endpoints, which would _not_ generate a "backendRef doesn't exist" error, but would end up with the same behavior).

evankanderson replied:
You might also get a `BackendRefEmpty` condition...

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+
+## Non-Goals
+
+
```
```suggestion
* Define the full set of Conditions that will ever be used with Gateway API
```

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+
+Gateway API currently has a lot of issues related to status, especially that
+status is inconsistent ([#1111][1111]), that names are hard to understand ([#1110][1110]),
+Reasons aren't explained properly ([#1362][1362]), and so on.
```
Tiny nit (feel free to ignore):
```suggestion
and that Reasons aren't explained properly ([#1362][1362]).
```

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+This document lays out large-scale changes to the way that we talk about resources,
+and the Conditions to match them. This means that there will be an unavoidable break
+in what constitutes a healthy or unhealthy resource, and there will be changes
+required for all conformant implementations to stay conformant.
```
Maybe specifically mention the concept of being conformant with a specific release of the API?
```suggestion
required for all implementations to be conformant with the release that includes these changes.
```

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+  field was when the controller last saw a resource.
+* Conditions shoud describe the _current state_ of the resource at observation
+time, which means that they should be an adjective (like `Ready`), or a past-tense
+verb (like `Attached`).
```
Would `Attaching` ever be valid as a condition? Maybe just verb or adjective is sufficient here?

youngnick replied:
That sort of usage is specifically called out as not desirable in the API conventions, and I agree. Conditions are supposed to be level-triggered, not edge-triggered, so having a condition that says that something is happening is a bit weird. (Should `Attaching` transition to `false` once the attaching process is complete?)

shaneutt replied:
For posterity, can we cite some sources for that assertion? It would probably actually be good as a part of this status effort to document somewhere _in the code_ that we're avoiding this due to conventions, and then point to those conventions, as this will help future contributors to more quickly avoid pursuing such things (or give them a path to who they need to convince in the greater kubernetes community).

youngnick replied:
This is almost a direct quote from the API conventions as it is, should I put a link in here to the exact line?

youngnick replied:
The exact quote is as follows, I can just inline if it's better, but I was trying to avoid just pasting in the entire section:

---

Condition type names should describe the current observed state of the resource, rather than describing the current state transitions. This typically means that the name should be an adjective ("Ready", "OutOfDisk") or a past-tense verb ("Succeeded", "Failed") rather than a present-tense verb ("Deploying"). Intermediate states may be indicated by setting the status of the condition to Unknown.

For state transitions which take a long period of time (e.g. more than 1 minute), it is reasonable to treat the transition itself as an observed state. In these cases, the Condition (such as "Resizing") itself should not be transient, and should instead be signalled using the True/False/Unknown pattern. This allows other observers to determine the last update from the controller, whether successful or failed. In cases where the state transition is unable to complete and continued reconciliation is not feasible, the Reason and Message should be used to indicate that the transition failed.

---

I don't believe that we generally have any process that we want to handle in the way called out in the exception bullet point. Happy to be guided here on what you'd like to see.

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+
+Because this GEP is mainly concerned with updating the Conditions we are setting in
+Gateway API resources' `status`, it's worth reviewing some important points about
+Conditions. (Thi information is mainly taken from the [Typical status properties][typstatus]
```
```suggestion
Conditions. (This information is mainly taken from the [Typical status properties][typstatus]
```

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+really viable:
+* All conditions must be negative polarity
+* All conditions must be positive polarity
+* Some conditions can be positive polarity, but most should be negative.
```
My own personal preference would be for conditions with a positive polarity to represent the normal healthy state and always be present, and for conditions with a negative polarity to represent very specific error states and to only be present when true. I think we should do everything we can to avoid the presence of double negatives like "Detached == False", but recognize that there are some error states that are best represented with a negative condition. I just don't think those negative conditions should be present when false.

youngnick replied:
If we're going to have a mixed-polarity set of conditions, then I think we need to define:
- which Conditions are positive polarity (Probably `Attached`, `Programmed`, `Ready`, and `Valid` if we do it)
- which Conditions are positive polarity and should always be present (Probably `Attached`, `Programmed`, and `Valid` if we do it).
- which Conditions are core (currently, everything other than `Ready`)
- which Conditions are extended (only `Ready`).
- which Conditions are negative (all the error conditions).

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+absence of `Error: true` implies everything's okay).
+
+Cons:
+* Double negatives in the happy case: `NoError: false`
```
This seems like a con of "All Conditions Positive" but not sure it applies when we allow both positive and negative conditions?

youngnick replied:
Ah yes, you're right.

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+
+### Should conditions always be added?
+
+Since we are mandating that conditions should always be positive, we should also
+specify that controllers always place all conditions on each resource when they
+reconcile it, even if that update sets the status to `unknown` because of
+information not being available.
```
I think this approach would lead to a _very_ noisy conditions list. I'd personally prefer leaving room for negative/error conditions to be unset unless they were true.

youngnick replied:
There's only one struct allowed for Conditions, so we can't change nil-values. What we'll need to say is "For all conditions other than <this list>, they must be negative polarity, and must only be present if they are status: true". Conversely, this also means that the absent value for these Condition types is "False", not "I don't know yet". Not the end of the world, but a little different to the API conventions.

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+
+For many implementations (certainly for Envoy-based ones), getting this information
+correctly and avoiding races on applying it is surprisingly difficult. For this
+reason, this GEP proposes we give the `Ready` condition a short window
```
I think I'd rather reserve `Ready` for "Ready at this moment" and instead use a different term to represent what we're actually describing - this config has been written to the data plane but we're not sure how long that will take to propagate. Previously `Reconciled` has been proposed to communicate this.

youngnick replied:
Okay, I'll update this section to:
- add `Ready` as an Extended support condition, that, if present, indicates that the data plane is fully configured. But because it's not always present, we'll need `Reconciled` to indicate what we have here - that everything is okay and the data plane will be ready soon.

robscott reviewed (commented):
Thanks for all the work on this @youngnick! This is a great start

youngnick commented:
Okay, to summarize the comments around condition naming and polarity:

We're going to end up with four possible positive polarity summary conditions:
- `ValidSpec` or similar; there aren't any issues in the config of _this resource on its own_
- `Attached`; there aren't any issues with the way this resources _attaches to others_
- `Programmed` or similar; the resource has been processed and the data plane is in the process of being configured. It should be ready "soon", where "soon" is implementation-specific.
- `Ready`; an optional, *Extended* condition that is only added when the implementation can guarantee that all required configuration has been accepted by the data plane and is running. Because it's Extended, it will have conformance tests to guarantee this behavior.

Other error conditions may be added that are negative polarity - the only thing relevant about these Conditions is if they are `status: true`. If they are not `status: true`, they should not be present.

Three or four summary, positive-polarity conditions seems on the high side to me, but if everyone feels that the additional clarity is worth the overhead of managing more Conditions.

I'm kind of inclined to say that anything we would use a `ValidSpec` summary condition for should probably be in the validation rules - whether they are in the OpenAPI definitions or the validating webhook doesn't really matter. I don't think the cases in which `ValidSpec` is True and `Attached` is false are going to be very common, so I don't really see the utility of that condition as much, personally. But if we really want it, we can do it.

youngnick commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+* In general, resources should be considered `Attached` if their config will
+generate some config in the underlying data plane.
+* All Conditions will be positive polarity (`GoodState: true`, not `BadState: false`)
+* All relevant Conditions for a resource must be added when it's observed.
```
So, does this mean that all conditions that have Core conformance should always be present? Or does this mean that a set of positive-polarity summary endpoints we name should always be present? The former could include any negative-polarity conditions we define as having Core conformance (which would clash with the "don't have double negatives" guidance we're looking for in other places.)

youngnick replied:
I've guessed "summary conditions" here - meaning just the positive polarity ones.

youngnick commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,328 @@
+in the healthy case is much better rules out the first option, so we are left to
+decide between enforcing that all conditions are positive, or that we have a mix.
+
+Having an arbitrary mix will make doing machine-based extraction of information
```
It means that anything consuming the Conditions will need to have a hard-coded list of the positive polarity conditions. Not a deal breaker, but it's another obstacle in the way of building a generic condition-parsing library.

howardjohn commented:
ValidSpec can have things not feasible to check in a webhook like temporary dependency issues (referenced service doesn't exist) or implementation specific issues (parent has FooGateway, which doesn't support the Bar field), so I think it holds weight.

If we do want to collapse we could have a single `Valid` field with reasons like `NotValid`, `NotAttached`, etc (naming needs work) possibly...?

Also re-iterating my comment on https://github.com/kubernetes-sigs/gateway-api/pull/1383#discussion_r967280691 since I think I wasn't quite clear - I think making a `Ready` optional condition is actually harmful. It seems like the type of field only useful if its required, and I am not sure how feasible it is to require it.

youngnick commented:
I guess in my mind, `ValidSpec` is always a requirement for `Attached`, so there's less added value in having the two bits of information. For me, in the cases you give:
* Service doesn't exist  would cause a resource either to program 500s into the data plane and add a more specific condition for HTTPRoute (the already existing `ResolvedRefs`), or cause another Route type that doesn't have an error code like HTTP to become unattached
* The FooGateway not supporting the Bar field would cause the HTTPRoute to fail to attach, and so `Attached` would be false anyway, with Reason and Message to indicate the reason.

To put this another way, for me, `ValidSpec` is _always_ a requirement for `Attached`, so the marginal utility of having both seems low. Invalid objects can't attach, and the reason they can't is that they're invalid.

Is it just the naming that's the problem, or is it the pivoting the status around the idea of attachment?

howardjohn commented:
> I guess in my mind, `ValidSpec` is always a requirement for `Attached`, so there's less added value in having the two bits of information. For me, in the cases you give:
> 
> * Service doesn't exist  would cause a resource either to program 500s into the data plane and add a more specific condition for HTTPRoute (the already existing `ResolvedRefs`), or cause another Route type - that doesn't have an error code like HTTP - to become unattached.

I think I am confused. I thought ValidSpec meant it was 100% valid. So we would make ValidSpec=false if there is a unknown service. I thought the proposal was to have *only* the conditions listed above, so how would `ResolvedRefs` be used?

> * The FooGateway not supporting the Bar field would cause the HTTPRoute to fail to attach, and so `Attached` would be false anyway, with Reason and Message to indicate the reason.

Maybe.. unless the field is "fail open" and we just ignore it...? I am not sure we have a clear spec here on unsupported fields. It would be a bummer to add a filter and then suddenly have an outage because it wasn't supported.

costinm commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,389 @@
+
+This document lays out large-scale changes to the way that we talk about resources,
+and the Conditions to match them. This means that there will be an unavoidable break
+in what constitutes a healthy or unhealthy resource, and there will be changes
```
Given the beta status of the API - I would keep the old behavior in place and not break anything, maybe until v1.
Implementations generating both old and new is a pain - but far less than impacting users.

youngnick replied:
I added a section about this just below.

costinm commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,389 @@
+
+The guidance about Conditions being added as soon as a controller sees a resource
+is a bit unclear - as written in the conventions, it seems to imply that _all_ 
+relevant conditions should always be added, even if their status has to be set to
```
I think it is reasonable for the controller to do this - as an indication the resource is in use by the controller.

It becomes problematic if a resource may be processed by multiple controllers - for example the discussion
of HttpRoute applying to both mesh and Gateway can only work if the type has a prefix (MeshReady). 

In general it would be good to indicate how to handle resources where multiple controllers operate on. While 
it's good to know which controller mess with the resource - maybe the controller should skip this step if the
resource is not ignored ( a mesh controller will ignore any route that don't target both mesh and gateway as example)

youngnick replied:
In most cases, the Conditions slice is in a separate struct that's part of a slice that's namespaced by controller name - so that controllers should be able to update only their own Conditions. I'd expect that mesh controllers should have a distinct controller name string, so this method should continue working.

robscott replied:
I think our practice of including `controllerName` for each distinct set of conditions should largely allow multiple controllers to cooperate on status for the same resource, but open to other potential improvements.

evankanderson replied:
How do `controllerName` conditions roll up to `Accepted`?

Also, controllers probably need to be sure to use PATCH on the `status` sub-resource or do conditional PUTs and avoid extra writes if the conditions haven't changed.

youngnick replied:
For some resources, all conditions are namespaced by `controllerName`. I agree we need to ensure that the rollup behaivor is specified. I'm going to add a section about partial validity that should help.

costinm commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,389 @@
+### Proposed changes summary
+
+* All the current Conditions that indicate that the resource is okay and ready
+for processing witll be replaced with `Attached`, except for GatewayClass (since
```
Sound good - but I think Gateway may need additional info, like Ready - since the process may involve 2 steps.
For example Attached will be set when the controller generates the configs - but Ready when certificates are provisioned. CertManager would watch for Attached and start the ACME process, create the Secret - and the gateway controller would switch to Ready.

youngnick replied:
I'm proposing a `Programmed` condition to summarize this without the "this is ready right now" implications of Ready. I'll put those changes into this summary.

candita commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,389 @@
+That is, the proposal is to replace:
+* `Scheduled` on Gateway
+* `Detached` on Listener
+* `Accpeted` on Route
```
```suggestion
* `Accepted` on Route
```

youngnick replied:
Thanks!

youngnick commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,389 @@
+
+SIG-API Machinery has been reluctant to add fields that would aid in machine-parsing
+of Conditions, especially fields that would indicate the polarity, because they
+are intended more for human consumption than machine consumption.
```
I think it's more that API Machinery has said specifically that they won't add a field for indicating the polarity of a Condition to the Condition struct. This was in a discussion raised by @evankanderson, but I can't seem to find the reference now.

youngnick commented:
Okay, I think I've captured all the outstanding comments in that update. Please take a look, our timebox for this last pass of reviews is in just under 48 hours, so that I can start working on the implementation.

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+### Proposed changes summary
+
+* All the current Conditions that indicate that the resource is okay and ready
+for processing witll be replaced with `Attached`, except for GatewayClass (since
```
```suggestion
for processing will be replaced with `Attached`, except for GatewayClass (since
```

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+* The summary conditions will be added into the CRD definitions
+as default values for `status.Conditions`, with a status of `Unknown`.
+* Resolving a comment that came up, documentation will be added to clarify that
+it's okay to add your own Conditions, and that implementations should namespace
```
We should probably also recommend that implementations publish both new and old conditions for several months to provide a smoother transition.

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+    means that the controller is unable to determine the status for some reason.
+    * `reason` is a CamelCase string that is a brief description of the reason why
+    the `status` is set the way it is.
+    * `observedGeneration` is an optional field that sets what the `metadata.generation`
+    field was when the controller last saw a resource.
```
I think we'd had some conversations about making this required for conformant Gateway implementations. Probably out of scope for this GEP, just wanted to make sure it was intentionally excluded.

youngnick replied:
Good point, I updated.

robscott reviewed (commented):
Thanks @youngnick! These are great updates. A few nits and a question about naming but really like this direction.

skriss commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+
+* All the current Conditions that indicate that the resource is okay and ready
+for processing witll be replaced with `Attached`, except for GatewayClass (since
+it is the root of the resource tree, it will stay with `Accepted`).
```
I had the same thought and am +1 to this, `Accepted` seems a little bit easier to understand for users IMO and can be consistently used across all resources.

howardjohn commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+* The resource is valid enough to produce some configuration in the underlying
+data plane.
+
+For Gateway, `Attached` also subsumes the functions of `Scheduled`: `Attached`
```
On some impls, provisioning takes about a minute. That may mean that some errors that are immediately known are masked. Or I suppose they aren't masked, but we need to wait the entire minute to know if its valid. The message field could say "all fields valid, waiting for deployment" perhaps, but that isn't in the machine parse-able section.

That may not warrant change, just a thought

youngnick replied:
I worded this carefully, so as not to imply that the provisioning needs to have _completed_, just that _it can be started_. The completion of infra provisioning should be handled by `Programmed` or `Ready`, or both.

howardjohn reviewed (commented):
I am not expert here things seem reasonable to me

evankanderson commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+* In general, resources should be considered `Attached` if their config is valid
+enough to generate some config in the underlying data plane.
+* There will be a limited set of positive polarity summary conditions, and a number
+of other specific negative-polarity error conditions.
```
This will make UI display somewhat harder (green dots / checkmarks / exclamation points), but I realize it may align with human reading a bit more easily.

I don't have a strong opinion here, just want to highlight the cost elsewhere in the stack.

robscott commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+[Typical status properties][typstatus] section of the guidelines.
+5. Conditions should be applied to a resource the first time the controller sees
+the resource. This seems to imply that _all conditions should be present on every
+resource owned by a controller_, but the rest of the conventions don't make this
```
I agree that we should have a set of permanent and reliable conditions that are always present, but I personally don't think it makes sense for error conditions to always be present, especially since that tends to lead to confusing double negative states.

sunjayBhatia commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+
+All of these conditions share the following meanings:
+* The resource has been accepted for processing by the controller
+* The resource is syntatically and semantically valid, and internally consistent
```
nit

```suggestion
* The resource is syntactically and semantically valid, and internally consistent
```

youngnick commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+of other specific negative-polarity error conditions.
+* All relevant positive-polarity summary Conditions for a resource must be added
+when it's observed.
+For example, HTTPRoutes must always have `Accepted` and `ResolvedRefs`, regardless
```
I hadn't updated, but with the change to `Accepted` everywhere, this is now correct, thanks.

youngnick commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+introduced. This means that the implementation has seen the config, has everything
+it needs, parsed it, and sent configuration off to the data plane. The configuration
+should be available "soon". We'll leave "soon" undefined for now.
+* The summary conditions will be added into the CRD definitions
+as default values for `status.Conditions`, with a status of `Unknown`.
```
I was intending to make it that the non-`controllerName` Conditions will be present, but I'll leave this out for now.

youngnick commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+GatewayClass resource (because it's the root of the resource tree and doesn't
+attach to anything).
+
+That is, the proposal is to replace:
+* `Scheduled` on Gateway
+* `Detached` on Listener
+* `Accepted` on Route
+
+with `Attached` in all these locations.
+
+GatewayClass will maintain the `Accepted` condition.
```
Yeah, agreed. I've migrated the proposal to this.

youngnick commented on site-src/geps/gep-1364.md:
```diff
@@ -0,0 +1,409 @@
+
+Positive polarity Conditions that describe the desirable state of the object must
+always be set. These are currently `Attached`, `ResolvedRefs`, and `Programmed`.
+Implementations that use `Ready` must also add it immediately, set to `false`.
```
Yeah, that seems like a good idea.

youngnick commented:
Okay, updated again, this feels like it's getting closer.

Changelog this time:
- Changed references to `Attached` to `Accepted`. I think @robscott and @skriss are right here, this is probably clearer.
- Added a section on partial validity to explain this better.
- Added the example @sunjayBhatia gave to the examples.

robscott commented:
Thanks @youngnick! This is a really well written GEP. Since previous generations of this PR have already been approved a couple times, I think we've got pretty clear consensus around this approach.

/lgtm

mikemorris reviewed (commented):
Quite late in doing a thorough review of this, but really happy with where it ended up - feels like a good step forwards in resolving some of the current ambiguity in status conditions. Thanks @youngnick!

robscott commented:
@mikemorris these are great bits of feedback, do you have time to create a follow up PR or issue to address any that are straightforward to fix? Concerned that otherwise we'll lose track of these.
